Update documentation for transition to quarterly releases

youngnick opened the pull request:
This one should go in after #4021.

Signed-off-by: Nick Young <ynick@vmware.com>

stevesloka reviewed (commented):
Just a couple nits of personal preference overall. 

nit: I'd take out the "we" bits through-out and make it more of the project is doing x and contour is released like this or that. 

> Contour is in the process of changing both to quarterly releases and three supported releases.

nit: I'd also change this to be present tense, we designed it and talked about it, so this is what we're doing.

youngnick commented:
Thanks @stevesloka

> nit: I'd take out the "we" bits through-out and make it more of the project is doing x and contour is released like this or that.

Fair enough, it will sound better that way.

>>Contour is in the process of changing both to quarterly releases and three supported releases.

> nit: I'd also change this to be present tense, we designed it and talked about it, so this is what we're doing.

I thought it's already in present tense? We are changing to quarterly releases and three supported releases. We haven't finished that yet though.

youngnick commented:
I've updated the language a bit per @stevesloka's suggestion. If this is okay, please feel free to merge.

stevesloka commented on site/content/resources/support.md:
```diff
@@ -5,20 +5,51 @@ layout: page
+## Supported releases
 
-Only the latest stable release is supported.
+Contour is changing both to quarterly releases and three supported releases.
```
```suggestion
Contour is releases quarterly and supports three versions starting with Contour v1.20.
```

sunjayBhatia replied:
maybe remove `is` in the suggestion as well

stevesloka commented on site/content/resources/support.md:
```diff
@@ -5,20 +5,51 @@ layout: page
-Only the latest stable release is supported.
+Contour is changing both to quarterly releases and three supported releases.
 
+The first Contour version covered by the quarterly release cadence will be Contour v1.20, scheduled for late October 2021.
```
```suggestion

```

youngnick replied:
I don't understand why you want to remove this? Right now, we need to ensure that during this transition period, this is clearly explained for people who are using Contour. That's the point of all of this present-tense language. This will not live forever, once we are fully moved over to the new release schedule, then I will update this again.

As of 1.19, we are *not* releasing quarterly or supporting three versions. Saying otherwise is incorrect, and confusing.

stevesloka commented on site/content/resources/support.md:
```diff
@@ -5,20 +5,51 @@ layout: page
 
+The first Contour version covered by the quarterly release cadence will be Contour v1.20, scheduled for late October 2021.
+
+At the time it is released, it will be the only supported version, and versions 1.21 and 1.22 will continue supporting back to Contour 1.20.
```
```suggestion

```

stevesloka commented on site/content/resources/support.md:
```diff
@@ -5,20 +5,51 @@ layout: page
 `:latest` is an alias for {{< param latest_version >}} which is the current stable release.
 
-When required we may release a patch release to address security issues, serious problems with no suitable workaround, or documentation issues.
+When required the project may release a patch release to address security issues, serious problems with no suitable workaround, or documentation issues.
```
What does this sentence mean, it's not clear to me what it's trying to tell the user. Just that we're going to, at any time, release a patch and that moves `:latest`?

youngnick replied:
Yes. This is a *policy*. If we don't say exactly what `:latest` means, it's possible to misunderstand.

stevesloka replied:
I guess the statement doesn't explain anything other than "we'll release something". I think what it's getting at is that the `:latest` tag will move at any time?

stevesloka commented on site/content/resources/support.md:
```diff
@@ -5,20 +5,51 @@ layout: page
+
+At the time it is released, it will be the only supported version, and versions 1.21 and 1.22 will continue supporting back to Contour 1.20.
+
+When Contour 1.23 releases (nine months later), Contour 1.20 will fall out of support.
```
```suggestion
When Contour 1.23 releases, Contour 1.20 will fall out of support.
```

stevesloka commented on site/content/resources/support.md:
```diff
@@ -5,20 +5,51 @@ layout: page
+When Contour 1.23 releases (nine months later), Contour 1.20 will fall out of support.
+
+The following table illustrates how this will work. The given dates are estimates, not guarantees.
+They are our best guess as to when each version will be released.
```
```suggestion

```

stevesloka commented on site/content/resources/support.md:
```diff
@@ -5,20 +5,51 @@ layout: page
+|---------|--------|-------|-------|-------|-------|
+|Q3 2021 (September 2021) | :heavy_check_mark: |
+|Q4 2021 (October 2021) | :negative_squared_cross_mark: | :heavy_check_mark: |
+|Q1 2022 (January 2021) | :negative_squared_cross_mark: | :heavy_check_mark: |:heavy_check_mark: |
```
```suggestion
|Q1 2022 (January 2022) | :negative_squared_cross_mark: | :heavy_check_mark: |:heavy_check_mark: |
```

stevesloka commented on site/content/resources/support.md:
```diff
@@ -5,20 +5,51 @@ layout: page
+|Q3 2021 (September 2021) | :heavy_check_mark: |
+|Q4 2021 (October 2021) | :negative_squared_cross_mark: | :heavy_check_mark: |
+|Q1 2022 (January 2021) | :negative_squared_cross_mark: | :heavy_check_mark: |:heavy_check_mark: |
+|Q2 2022 (April 2021) | :negative_squared_cross_mark: | :heavy_check_mark: |:heavy_check_mark: |:heavy_check_mark: |
```
```suggestion
|Q2 2022 (April 2022) | :negative_squared_cross_mark: | :heavy_check_mark: |:heavy_check_mark: |:heavy_check_mark: |
```

stevesloka commented on site/content/resources/support.md:
```diff
@@ -5,20 +5,51 @@ layout: page
+|Q4 2021 (October 2021) | :negative_squared_cross_mark: | :heavy_check_mark: |
+|Q1 2022 (January 2021) | :negative_squared_cross_mark: | :heavy_check_mark: |:heavy_check_mark: |
+|Q2 2022 (April 2021) | :negative_squared_cross_mark: | :heavy_check_mark: |:heavy_check_mark: |:heavy_check_mark: |
+|Q3 2022 (July 2021) | :negative_squared_cross_mark: | :negative_squared_cross_mark: |:heavy_check_mark: |:heavy_check_mark: | :heavy_check_mark: |
```
```suggestion
|Q3 2022 (July 2022) | :negative_squared_cross_mark: | :negative_squared_cross_mark: |:heavy_check_mark: |:heavy_check_mark: | :heavy_check_mark: |
```

stevesloka commented on site/content/docs/v1.18.1/github.md:
```diff
@@ -2,7 +2,7 @@ This document outlines how we use GitHub.
 ## Milestones
 
-Contour attempts to ship on a monthly-ish, generally every four to six weeks, basis.
+Contour attempts to ship on a quarterly basis.
```
```suggestion
Contour ships on a quarterly basis, about every three months from the last release.
```

stevesloka commented on site/content/docs/main/github.md:
```diff
@@ -2,7 +2,7 @@ This document outlines how we use GitHub.
 ## Milestones
 
-Contour attempts to ship on a monthly-ish, generally every four to six weeks, basis.
+Contour attempts to ship on a quarterly basis.
```
```suggestion
Contour ships on a quarterly basis, about every three months from the last release.
```

stevesloka commented on RELEASES.md:
```diff
@@ -2,17 +2,23 @@
-* [Contour Compatibility Matrix](https://projectcontour.io/resources/compatibility-matrix/)
+Per the [Contour support policy](https://projectcontour.io/resources/support/), the project is in the process of transitioning to supporting three Contour releases. Please see the support policy page to see what versions are currently supported.
+
+Also, please consult the [Contour Compatibility Matrix](https://projectcontour.io/resources/compatibility-matrix/) for details of what each version of Contour requires for each of its dependencies like Envoy, Kubernetes, and so on.
```
```suggestion
Consult the [Contour Compatibility Matrix](https://projectcontour.io/resources/compatibility-matrix/) for details of what each version of Contour requires for each of its dependencies like Envoy, Kubernetes, and so on.
```

stevesloka commented on RELEASES.md:
```diff
@@ -2,17 +2,23 @@
 
 ### Patch releases
-Patch releases are based on the major/minor release branch. There is no specific release cadence for patch releases as we already release monthly. However, we will create patch releases to address critical community and security issues (for example to address high severity security issues in Contour or in Envoy). We will patch release only the latest release of Contour. The Contour team only main ...
+Patch releases are based on the major/minor release branch. There is no specific release cadence for patch releases. However, the project will create patch releases to address critical community and security issues (for example to address high severity security issues in Contour or in Envoy).The project will issue patch releases for all supported versions of Contour.
```
```suggestion
Patch releases are based on the major/minor release branch. Patch releases will be created to address critical community and security issues (for example to address high severity security issues in Contour or in Envoy)and will apply to all supported versions of Contour.
```

stevesloka commented on RELEASES.md:
```diff
@@ -2,17 +2,23 @@
 
 ### Major and Minor Releases
-Major and minor releases of Contour will be branched from main when the release reaches to `RC (release candidate)` state. The release cadence is around once a month, If for any reason this release cadence has to be adjusted (for example due to open source events), we will communicate it clearly on Slack, Twitter, and distribution lists. There is no mandated timeline for major versions and there  ...
+Major and minor releases of Contour will be branched from main when the release reaches to `RC (release candidate)` state. The release cadence is currently once a month, but is migrating to quarterly as of the October 2021 release (which will be Contour 1.20).
```
```suggestion
Major and minor releases of Contour will be branched from the `main` branch when the release reaches `RC (release candidate)` state. 
```

stevesloka commented on RELEASES.md:
```diff
@@ -2,17 +2,23 @@
 
 ## Releases
-Contour releases will be versioned using dotted triples, similar to [Semantic Version](http://semver.org/). For this specific document, we will refer to the respective components of this triple as `<major>.<minor>.<patch>`. The version number may have additional information, such as "-rc1,-rc2,-rc3" to mark release candidate builds for earlier access. Such releases will be considered as "pre-rele ...
+Contour releases will be versioned using dotted triples, similar to [Semantic Version](http://semver.org/). The project refers to the respective components of this triple as `<major>.<minor>.<patch>`. The version number may have additional information, such as "-rc1,-rc2,-rc3" to mark release candidate builds for earlier access. Such releases will be considered as "pre-releases".
```
```suggestion
Contour releases will be versioned using dotted triples, similar to [Semantic Version](http://semver.org/) with  respective components of this triple as `<major>.<minor>.<patch>`. The version number may have additional information, such as "-rc1,-rc2,-rc3" to mark release candidate builds for earlier access. Such releases will be considered as "pre-releases".
```

stevesloka commented on RELEASES.md:
```diff
@@ -2,17 +2,23 @@
+
+Also, please consult the [Contour Compatibility Matrix](https://projectcontour.io/resources/compatibility-matrix/) for details of what each version of Contour requires for each of its dependencies like Envoy, Kubernetes, and so on.
+
+Both of these details are also available in a machine-consumable (YAML) format via [versions.yaml](https://github.com/projectcontour/contour/blob/main/versions.yaml).
```
```suggestion
Both of these details are also available in a machine-consumable (YAML) format via [versions.yaml](https://raw.githubusercontent.com/projectcontour/contour/main/versions.yaml).
```

xaleeks commented on site/content/resources/support.md:
```diff
@@ -5,20 +5,51 @@ layout: page
+- If you are not running the supported version from your release train, you'll need to upgrade first if you have any problems.
+- When a new patch is cut, that will become the supported version for that release.
+
+So, if Contour 1.20.0 is the only supported version, and Contour 1.20.1 is released, then the supported version will change to 1.20.1.
```
wait, are you saying we will only support the latest patch of a supported minor?  So if 1.20 is a supported minor with 5 patches out, only 1.20.5 is supported?

youngnick replied:
Yes. If not, then we have no way to fix problems. If there is a problem with 1.20.4, and we cut 1.20.5, how else do we say "to fix this problem, you have to upgrade"? This is pretty standard tbh.

xaleeks commented on site/content/resources/support.md:
```diff
@@ -5,20 +5,51 @@ layout: page
+
+So, if Contour 1.20.0 is the only supported version, and Contour 1.20.1 is released, then the supported version will change to 1.20.1.
+
+If, later on, the supported versions are 1.21, 1.22, and 1.23, and 1.21.1, 1.22.1, and 1.23.1 are released, then the patch releases will be the only supported versions.
```
so 1.21.0, 1.22.0, 1.23.0 will fall out of support?

youngnick replied:
Yes. Is this not a standard feature of a patch release?

xaleeks replied:
In thinking about this again, I don't think there's a difference btw the two so my question is moot I suppose.  The question I had in mind originally is if our stance is such that 1.20.5 comes out and renders anything <= 1.20.4 out of support, then what happens if another vulnerability is found on, say, 1.20.2. Presumably that same vulnerability still exists on the latest 1.20.5 because the delta btw 1.20.5 and the previous 1.20.4 is precisely some patched vulnerability.  So we'd be forced to release a 1.20.6.

Now if for some reason 1.20.5 was not prone to this vulnerability, I think we'd be fine recommending users to just upgrade to 1.20.5.

youngnick replied:
Yes, that's how I expected everything to work too.

youngnick commented on site/content/resources/support.md:
```diff
@@ -5,20 +5,51 @@ layout: page
+
+When Contour 1.23 releases (nine months later), Contour 1.20 will fall out of support.
+
+The following table illustrates how this will work. The given dates are estimates, not guarantees.
```
If we don't say that the dates are estimates, then people can say "wait, you didn't release when you said you would".

youngnick commented:
@stevesloka, I don't understand what all your suggestions are trying to achieve. Can we talk about this in the maintainer meeting tomorrow?

stevesloka commented:
It's my understanding that this doc explains how we are going to release, not when. Showing examples of how things are supported are fine, but as we all know, schedules change and this will quickly get out of date. 

We should be explaining our release schedule with milestones (https://github.com/orgs/projectcontour/projects/2) which allows us to groom the issues+features and updates those dates accordingly.

youngnick commented:
Thanks for the feedback @stevesloka, I generally agree, but in this case the transition is going to take a long time (nearly a year), so I want to have something that covers us in the meantime.

I'd like to get this in, and I will take on ensuring that it's updated. Does that sound okay?

stevesloka commented:
@youngnick I don't agree that this needs to have exact dates. I still think it should only cover the **process** and not when we're releasing. 

We should have a good way to show what the releases look like as we move along with the project board. As soon as we miss those dates then this document contradicts the project board which contradicts the milestones and all needs to be kept tightly in sync. Users then have less information since nothing aligns.

That said, this is just my 2 cents, so I'll let @sunjayBhatia, @skriss, @xaleeks, and @tsaarni add their thoughts.

sunjayBhatia reviewed (approved):
general content etc. lgtm

maybe a solution to the discussion around dates would be to use year + quarter rather than months? Nick mentioned we would need to proactively keep this up to date as things change but that could maybe reduce some churn needed?

tsaarni reviewed (approved):
LGTM.
I don't have strict opinion regarding the months being listed or not, but as a user, I might be bit intimidated by the project board and would maybe prefer a table like proposed here. On the other hand, in the current website, I'm not sure if I would necessarily bump into the support page if I'm looking for release schedule? For understanding support strategy only, the table without dates or table with quarters only could suffice as well.

youngnick commented:
Thanks for the feedback @sunjayBhatia and @tsaarni. I've removed the months and added a short section to the "Tagging" page about adding a `:v1.xx` that we'll keep up to date with the latest patch.
